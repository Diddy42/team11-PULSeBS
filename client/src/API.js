async function userLogin(username, password) {
    return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/login.php', {     //Da verificare
            method: 'POST',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type' : 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({username: username, password: password}),
        }).then((response) => {
            response.json().then((obj) => { 
                console.log(obj);
                if(obj === 0 ) {        //Da modificare in base a quanto ritorna il server. Sarebbe utile che ritorni l'id o anche nome e cognome la visualizzare, e il ruolo
                    reject(obj);
                } else {
                    resolve(obj);    
                }
            })
            .catch((err) => { reject({ errors: [{ param: "Application", msg: "Cannot parse server response" }] }) });
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function logout(){
	return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/logout.php')      //Da verificare
        .then((response) => {
			if(response.ok){
				resolve();
			}
			else{
				reject();
			}
       })
       .catch((err) => { 
      		 reject(err); 
       });
    });
}


async function getStudentLectures(studentId) {
    const url = "test"                  //Da definire
    const response = await fetch(url);
    const lectures = await response.json();
    if(response.ok) {
        return lectures;
    } else {
        let err = {status: response.status, errorObj: lectures};
        throw err; 
    }
}

async function getBooking(lectureId) {
    const url = "test"                  //Da definire
    const response = await fetch(url);
    const booking = await response.json();
    if(response.ok) {
        return booking;
    } else {
        let err = {status: response.status, errorObj: booking};
        throw err; 
    }
}

async function bookASeat(lectureId, studentId) {
    const url = "test"                  //Da definire
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type' : 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({lectureId: lectureId, studentId: studentId}),
    });
    
    try{
        const resJ = await response.json();
        if(response.ok) {
		    return resJ;
		} else {
		    let err = {status: response.status, errorObj: resJ};
		    throw err; 
		}
    }
    catch(e){
        console.log("Error in booking a seat: " + e);
        return 0;   //meaning: no new customer
    }
}

const API = {
    userLogin, logout, getStudentLectures, getBooking, bookASeat
};
export default API;
