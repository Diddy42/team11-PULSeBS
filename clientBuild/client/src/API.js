async function userLogin(username, password) {
    return new Promise((resolve, reject) => {
        fetch('http://localhost/server/src/api/Login.php', {
            method: 'POST',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({ username: username, password: password }),
        }).then((response) => {
            response.json().then((obj) => {
                /*
                if(obj === 0 ) {        //Da modificare in base a quanto ritorna il server. Sarebbe utile che ritorni l'id o anche nome e cognome la visualizzare, e il ruolo
                    reject(obj);
                } else {
                    resolve(obj);    
                }
                */
                resolve(obj);
            })
                .catch((err) => { reject({ errors: [{ param: "Application", msg: "Cannot parse server response" }] }) });
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function logout() {
    return new Promise((resolve, reject) => {
        fetch('http://localhost:80/server/api/logout.php')      //Da verificare
            .then((response) => {
                if (response.ok) {
                    resolve(0);
                }
                else {
                    reject();
                }
            })
            .catch((err) => {
                reject(err);
            });
    });
}

//Ritorna le lezioni prenotabili dallo studente
async function getBookableStudentLectures(studentId) {
    const url = "http://localhost/server/bookableLessons/" + studentId;
    const response = await fetch(url);
    const lectures = await response.json();
    if (response.ok) {
        if (lectures === 0) {
            return [];
        }
        return lectures;
    } else {
        let err = { status: response.status, errorObj: lectures };
        throw err;
    }
}

//Ritorna le prenotazioni effettuate dagli studenti per la lezione indicata (teacher side)
async function getBooking(lectureId) {
    const url = "http://localhost/server/bookedStudentsForLecture/" + lectureId;
    const response = await fetch(url);
    const booking = await response.json();
    if (response.ok) {
        if (booking === 0) {
            return [];
        }
        return booking;
    } else {
        let err = { status: response.status, errorObj: booking };
        throw err;
    }
}

// API to retrieve all the bookings of a student
async function getStudentBookings(studentId) {
    const url = "http://localhost/server/studentBookings/" + studentId;
    const response = await fetch(url);
    const booking = await response.json();
    if (response.ok) {
        if (booking === 0) {
            return [];
        }
        return booking;
    } else {
        let err = { status: response.status, errorObj: booking };
        throw err;
    }
}

//API per prenotare un posto a lezione
async function bookASeat(lectureId, studentId, date) {
    const url = "http://localhost/server/insertLecture"
    const response = await fetch(url, {
        method: 'POST',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
            idLesson: lectureId,
            idUser: studentId,
            date: date
        }),
    });

    try {
        const resJ = await response.json();
        if (response.ok) {
            return resJ;
        } else {
            let err = { status: response.status, errorObj: resJ };
            throw err;
        }
    }
    catch (e) {
        console.log("Error in booking a seat: " + e);
        return 0;   //meaning: no new customer
    }
}

// API to retrieve all the lectures of a teacher
async function getTeacherLectures(teacherId) {
    const url = "http://localhost/server/scheduledLecturesForTeacher/" + teacherId
    const response = await fetch(url);
    const lectures = await response.json();
    if (response.ok) {
        if (lectures === 0) {
            return [];
        }
        return lectures;
    } else {
        let err = { status: response.status, errorObj: lectures };
        throw err;
    }
}

async function deleteBooking(bookingId) {
    const url = "http://localhost/server/updateBooking/" + bookingId;
    return new Promise((resolve, reject) => {
        fetch(url, {
            method: 'PUT',
            headers: {
                //'Content-Type': 'application/json',
                'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
            },
            body: JSON.stringify({
            }),
        }).then((response) => {
            if (response.ok) {
                resolve(0);
            } else {
                reject({ errore: "Error in deleting a booking" });
            }
        }).catch((err) => { reject({ errors: [{ param: "Server", msg: "Cannot communicate" }] }) });
    });
}

async function deleteLecture(lectureId) {
    const url = "http://localhost/server/deleteLecture/" + lectureId;
    const response = await fetch(url, {
        method: 'PUT',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
        }),
    });
    const result = await response.json();
    if (response.ok) {
        if (result === 1) {
            return 1;
        } else {
            console.log("Too late to delete the lecture");
            return -1;
        }
    } else {
        let err = { status: response.status, errorObj: result };
        throw err;
    }
}

async function changeToOnline(lectureId) {
    const url = "http://localhost/server/changeToOnline/" + lectureId;
    const response = await fetch(url, {
        method: 'PUT',
        headers: {
            //'Content-Type': 'application/json',
            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
        },
        body: JSON.stringify({
        }),
    });
    const result = await response.json();
    if (response.ok) {
        if (result === 1) {
            return 1;
        } else {
            console.log("Too late to change the lecture");
            return -1;
        }
    } else {
        let err = { status: response.status, errorObj: result };
        throw err;
    }
}

async function getAllCourses() {
    const url = "http://localhost/server/courses";

    const response = await fetch(url);
    const courses = await response.json();
    if (response.ok) {
        if (courses === 0) {
            return [];
        }
        return courses;
    } else {
        let err = { status: response.status, errorObj: courses };
        throw err;
    }
}

const ALL_COURSES_FILTER = "L.idCourse";
async function getBookingStatisticsByMonth(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=year,monthOfYear&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
} 
async function getBookingStatisticsByWeek(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=year_month_week&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
}
async function getBookingStatisticsByLesson(courseIds) {
    const url = `http://localhost/server/bookingStatistics/?filterTime=L.idLesson&filterCourse=${courseIds.toString()}`;

    const response = await fetch(url);
    const stats = await response.json();
    if (response.ok) {
        if (stats === 0) {
            return [];
        }
        return stats;
    } else {
        let err = { status: response.status, errorObj: stats };
        throw err;
    }
}

async function getCoursesOfTeacher(teacherId) {
    const url = "http://localhost/server/teacherCourses/" + teacherId;

    const response = await fetch(url);
    const courses = await response.json();
    if (response.ok) {
        if (courses === 0) {
            return [];
        }
        return courses;
    } else {
        let err = { status: response.status, errorObj: courses };
        throw err;
    }
}

const API = {
    userLogin, logout, getBookableStudentLectures, getBooking, getStudentBookings, bookASeat, deleteBooking, getTeacherLectures, deleteLecture, changeToOnline, getAllCourses, 
    getBookingStatisticsByMonth, getBookingStatisticsByWeek, getBookingStatisticsByLesson, ALL_COURSES_FILTER, getCoursesOfTeacher
};
export default API;
