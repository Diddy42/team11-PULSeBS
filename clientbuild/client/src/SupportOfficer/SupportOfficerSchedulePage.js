import React from "react";
import { Container, Row, Col, Button, Table, Form } from 'react-bootstrap';
import { Link } from "react-router-dom";
import API from '../API';

class SupportOfficerSchedulePage extends React.Component {
  constructor(props){
    super(props);

    this.state = { allCourses: [], filteredCourses: [] };

  }

  render(){
    return <>
      <Container fluid>
        <Row>
          <Col className='col-3'>
            <Form className='mt-3'>
                <Form.Group controlId="formCourseName">
                    <Form.Label>Filter courses by name or by teacher's name:</Form.Label>
                    <Form.Control type="text" placeholder="Software Engineering" onChange={this.handleOnChangeText}/>
                </Form.Group>
            </Form>
          </Col>

          <Col className='col-8'>
              <Table className='mt-3' striped bordered hover>
                  <thead>
                      <tr>
                          <th>Name</th>
                          <th>Id</th>
                          <th>Teacher</th>
                          <th>Action</th>
                      </tr>
                  </thead>

                  <tbody>
                      {this.state.filteredCourses.map(c => this.createCourseTableRow(c))}
                  </tbody>
              </Table>
          </Col>
        </Row>
      </Container>
    </>

  }

  componentDidMount = () => {
    API.getAllCourses().then((courses) => {
        this.setState({
            allCourses: courses,
            filteredCourses: courses
        });
    });

  }

  createCourseTableRow = (c) => {
      return <tr key={c.idCourse}>
          <td>{c.courseName}</td>
          <td>{c.idCourse}</td>
          <td>{c.teacherName}</td>
          <td>
              <Link to={"/support-officer/updateSchedule/" + c.idCourse}>
                <Button className='mt-1' variant='danger' onClick={() => {this.handleButtonClick(c);}}>
                  Change schedule
                </Button>
              </Link>
          </td>
      </tr>
  }
  
  handleButtonClick = (c) => {

  }

  courseMatchesInputTeacher = (c, wordsInInput) => {
      var res = true;
      var i;

      for(i = 0; i < wordsInInput.length; i++){
          if(!c.teacherName.toLowerCase().includes(wordsInInput[i])){
              res = false;
          }
      }

      return res;
  }

  courseMatchesInputName = (c, wordsInInput) => {
      var res = true;
      var i;

      for(i = 0; i < wordsInInput.length; i++){
          if(!c.courseName.toLowerCase().includes(wordsInInput[i])){
              res = false;
          }
      }

      return res;
  }

  courseMatchesInputId = (c, wordsInInput) => {
      var res = true;
      var i;

      for(i = 0; i < wordsInInput.length; i++){
          if(!c.idCourse.toLowerCase().includes(wordsInInput[i])){
              res = false;
          }
      }

      return res;
  }

  handleOnChangeText = (event) => {
      var currentTextInput = event.target.value.toLowerCase();
      var wordsInInput = currentTextInput.split(' ').filter(w => w !== '');

      var filteredCourses = this.state.allCourses.filter(
        c => (this.courseMatchesInputName(c, wordsInInput) || this.courseMatchesInputTeacher(c, wordsInInput) || this.courseMatchesInputId(c, wordsInInput))
      );
      this.setState({ filteredCourses: filteredCourses });
  }

}

export default SupportOfficerSchedulePage;
