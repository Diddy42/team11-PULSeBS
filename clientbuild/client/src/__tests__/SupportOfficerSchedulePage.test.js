import TestRenderer from 'react-test-renderer';
import React from 'react';
import SupportOfficerSchedulePage from '../SupportOfficer/SupportOfficerSchedulePage';
import API from '../API';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import { fireEvent } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

API.getAllCourses = jest.fn(() => Promise.resolve([{courseName: "test", idCourse: 1, teacherName: "Test Test"}]));
test('renders SupportOfficerSchedulePage page', () => {
    const tree = TestRenderer.create(<SupportOfficerSchedulePage/>).toJSON();
    expect(tree).toMatchSnapshot();
  });

test('courseMatchesInputId', async () => {
    const component = await shallow(<SupportOfficerSchedulePage/>);
    const instance = await component.instance();
    let res1 = instance.courseMatchesInputId({idCourse: "test"}, "re");
    expect(res1).toBe(false);
    let res2 = instance.courseMatchesInputId({idCourse: "test"}, "te");
    expect(res2).toBe(true);
});

test('courseMatchesInputName', async () => {
    const component = await shallow(<SupportOfficerSchedulePage/>);
    const instance = await component.instance();
    let res1 = instance.courseMatchesInputName({courseName: "test"}, "re");
    expect(res1).toBe(false);
    let res2 = instance.courseMatchesInputName({courseName: "test"}, "te");
    expect(res2).toBe(true);
});

test('courseMatchesInputTeacher', async () => {
    const component = await shallow(<SupportOfficerSchedulePage/>);
    const instance = await component.instance();
    let res1 = instance.courseMatchesInputTeacher({teacherName: "test"}, "re");
    expect(res1).toBe(false);
    let res2 = instance.courseMatchesInputTeacher({teacherName: "test"}, "te");
    expect(res2).toBe(true);
});

test('handleOnChangeText', async () => {
    const component = await shallow(<SupportOfficerSchedulePage/>);
    const instance = await component.instance();
    instance.setState({allCourses: [{"idCourse":"XY1211","courseName":"test","teacherName":"test"}]})
    instance.courseMatchesInputName = jest.fn(() => true);
    instance.courseMatchesInputTeacher = jest.fn(() => true);
    instance.courseMatchesInputId = jest.fn(() => true);
    await instance.handleOnChangeText({target: {value: "test sample"}});
    expect(instance.state.filteredCourses).toStrictEqual([{"idCourse":"XY1211","courseName":"test","teacherName":"test"}]);
    instance.courseMatchesInputName = jest.fn(() => false);
    instance.courseMatchesInputTeacher = jest.fn(() => false);
    instance.courseMatchesInputId = jest.fn(() => false);
    await instance.handleOnChangeText({target: {value: "test sample"}});
    expect(instance.state.filteredCourses).toStrictEqual([]);
});

test("handle button click", async () => {
    const component = await shallow(<SupportOfficerSchedulePage/>);
    const instance = await component.instance();
    instance.handleButtonClick = jest.fn();
    let button = component.find("#testButton1");
    button.simulate("click");
    expect(instance.handleButtonClick).toHaveBeenCalledTimes(1);
});