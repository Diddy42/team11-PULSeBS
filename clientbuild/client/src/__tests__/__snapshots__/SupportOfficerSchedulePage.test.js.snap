// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SupportOfficerSchedulePage page 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col-3 col"
    >
      <form
        className="mt-3"
      >
        <div
          className="form-group"
        >
          <label
            className="form-label"
            htmlFor="formCourseName"
          >
            Filter courses by name or by teacher's name:
          </label>
          <input
            className="form-control"
            id="formCourseName"
            onChange={[Function]}
            placeholder="Software Engineering"
            type="text"
          />
        </div>
      </form>
    </div>
    <div
      className="col-8 col"
    >
      <table
        className="mt-3 table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Id
            </th>
            <th>
              Teacher
            </th>
            <th>
              Action
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
</div>
`;
